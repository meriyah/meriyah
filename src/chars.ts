/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
