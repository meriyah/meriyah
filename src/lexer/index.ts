export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
