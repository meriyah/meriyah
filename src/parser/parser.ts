import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
