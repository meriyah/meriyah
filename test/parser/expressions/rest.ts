import * as t from 'node:assert/strict';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail, pass } from '../../test-utils';
describe('Expressions - Rest', () => {
  for (const arg of [
    'let { ...x = y } = z;',
    'let { a, ...b, c } = x;',
    'let {...obj1,...obj2} = foo',
    'let {...obj1,a} = foo',
    'let {...(obj)} = foo',
    '({...(a,b)}) => {}',
    'let {...obj1,} = foo',
    'let {...(a,b)} = foo',
    '({...x = 1} = {})',
    'var {...x = 1} = {}',
    'function test({...x = 1}) {}',
    '({...[]} = {})',
    '({...{}} = {})',
    '({...(obj)}) => {}',
    '({...(a,b)}) => {}',
    '({...[a,b]} = foo)',
    '({...{a,b}} = foo)',
    '({...(a,b)} = foo)',
    'let {...(a,b)} = foo',
    'let {...(obj)} = foo',
    'let {...obj1,...obj2} = foo',
    'let {...obj1,a} = foo',
    'let {...obj1,} = foo',
    'let { ...x = y } = z;',
    'let { a, ...b, c } = x;',
    // Object rest element needs to be the last AssignmenProperty in ObjectAssignmentPattern.
    '{...rest, b}',
    'function test({...[]}) {}',
    'var {...[]} = {}',
    'function test({...{a}}) {}',
    // Babylon PR: https://github.com/babel/babylon/issues/667
    ' ( {...{}} = {} ) ',
    // Babylon issue: https://github.com/babel/babylon/issues/661
    'let {...{}} = {};',
    '({...[a,b]}) => {}',
    '({...obj1,a} = foo)',
    '({...obj1,} = foo)',
    'let {...[a,b]} = foo',
    'let {...{a,b}} = foo',
  ]) {
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`);
      });
    });

    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { next: true });
      });
    });

    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { webcompat: true });
      });
    });
  }

  for (const arg of [
    '({...obj} = foo)',
    '({a,...obj} = foo)',
    '({a:b,...obj} = foo)',
    '({...obj}) => {}',
    '({...obj} = {}) => {}',
    '({a,...obj}) => {}',
    '({a:b,...obj}) => {}',
    '({...rest})',
    'let { x, y, ...z } = { x: 1, y: 2, a: 3, b: 4 };',
    'let { ...x } = y;',
    '({a, b, ...{c, e}})',
    '({ x, ...{y , z} })',
    'function f({ x, y, ...z }) {}',
    '({...(obj)} = foo)',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`);
      });
    });
  }

  for (const arg of ['function f(a, ...b, c) {}']) {
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`);
      });
    });
  }

  for (const arg of [
    'function empty(...{}) {}',
    'function emptyWithArray(...{p: []}) {}',
    'function emptyWithObject(...{p: {}}) {}',
    'function emptyWithLeading(x, ...{}) {}',
    'function singleElement(...{a: b}) {}',
    'function singleElementWithInitializer(...{a: b = 0}) {}',
    'function singleElementWithArray(...{p: [a]}) {}',
    'function singleElementWithObject(...{p: {a: b}}) {}',
    'function singleElementWithLeading(x, ...{a: b}) {}',
    'function multiElement(...{a: r, b: s, c: t}) {}',
    'function multiElementWithInitializer(...{a: r = 0, b: s, c: t = 1}) {}',
    'function multiElementWithArray(...{p: [a], b, q: [c]}) {}',
    'function multiElementWithObject(...{a: {p: q}, b: {r}, c: {s = 0}}) {}',
    'function multiElementWithLeading(x, y, ...{a: r, b: s, c: t}) {}',
    'function empty(...[]) {}',
    'function emptyWithArray(...[[]]) {}',
    'function emptyWithObject(...[{}]) {}',
    'function emptyWithRest(...[...[]]) {}',
    'function emptyWithLeading(x, ...[]) {}',
    'function singleElement(...[a]) {}',
    'function singleElementWithInitializer(...[a = 0]) {}',
    'function singleElementWithArray(...[[a]]) {}',
    'function singleElementWithObject(...[{p: q}]) {}',
    'function singleElementWithRest(...[...a]) {}',
    'function singleElementWithLeading(x, ...[a]) {}',
    'function multiElement(...[a, b, c]) {}',
    'function multiElementWithInitializer(...[a = 0, b, c = 1]) {}',
    'function multiElementWithArray(...[[a], b, [c]]) {}',
    'function multiElementWithObject(...[{p: q}, {r}, {s = 0}]) {}',
    'function multiElementWithRest(...[a, b, ...c]) {}',
    'function multiElementWithLeading(x, y, ...[a, b, c]) {}',
    'function af(...a) {}',
    'function bf(a, ...b) {}',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { webcompat: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true });
      });
    });
  }

  fail('Expressions - Rest (fail)', [
    'function foo(...[a], ...b) {}',
    'function foo(...a, ...[b]) {}',
    'function foo(a, ...b, c) => {}',
    'function foo(a, ...[b], c) => {}',
    'var obj = class { method(a, b = 1, ...c = [2,3]) {} };',
    'function f(a, ...b) { "use strict"; }',
    'function f(a, ...[b]) { "use strict"; }',
    'var x = { set setter(...[x]) {} }',
    'var x = class { set setter(...x) {} }',
    'var x = class { set setter(...[x]) {} }',
    '(a = ...NaN, b = [...[1,2,3]], ...rest) => {};',
    '(a = (...NaN), ...b = [...[1,2,3]], rest) => {};',
    '(a = [...NaN], ...b = [...[1,2,3]], rest) => {};',
    '(a, ...b, ...rest) => {};',
    '(...rest = ...NaN) => {};',
    '[...x,] = [1,2,3];',
    '[...x, y] = [1,2,3];',
    'function foo(...[a],) {}',
  ]);

  pass('Expressions - Rest (pass)', [
    'var obj = { method(a, b, c, ...[d]) { return [a, b, c, d]; } };',
    "function objRest(...{'0': a, '1': b, length}) { return [a, b, length]; }",
    'function singleRest(...[d]) { return d; }',
    'function foo(a, b, c, ...[d]) { arguments; return [a, b, c, d]; }',
    'class restClass { method(a, b, c, ...[d]) { arguments; return [a, b, c, d]; } };',
    'function fooInline(a, b, c, ...rest) { arguments; this; return [a, b, c, ...rest]; }',
    'var func5 = function (...[argArr13]) { function foo() { eval(); } };',
  ]);
});
