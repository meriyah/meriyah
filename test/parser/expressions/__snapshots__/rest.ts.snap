// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - Rest > Expressions - Rest (fail) > (...rest = ...NaN) => {}; 1`] = `
"SyntaxError [1:11-1:14]: Unexpected token: '...'
> 1 | (...rest = ...NaN) => {};
    |            ^^^ Unexpected token: '...'"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > (a = (...NaN), ...b = [...[1,2,3]], rest) => {}; 1`] = `
"SyntaxError [1:13-1:14]: Expected '=>'
> 1 | (a = (...NaN), ...b = [...[1,2,3]], rest) => {};
    |              ^ Expected '=>'"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > (a = ...NaN, b = [...[1,2,3]], ...rest) => {}; 1`] = `
"SyntaxError [1:5-1:8]: Unexpected token: '...'
> 1 | (a = ...NaN, b = [...[1,2,3]], ...rest) => {};
    |      ^^^ Unexpected token: '...'"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > (a = [...NaN], ...b = [...[1,2,3]], rest) => {}; 1`] = `
"SyntaxError [1:34-1:35]: Invalid rest argument
> 1 | (a = [...NaN], ...b = [...[1,2,3]], rest) => {};
    |                                   ^ Invalid rest argument"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > (a, ...b, ...rest) => {}; 1`] = `
"SyntaxError [1:8-1:9]: Invalid rest argument
> 1 | (a, ...b, ...rest) => {};
    |         ^ Invalid rest argument"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > [...x, y] = [1,2,3]; 1`] = `
"SyntaxError [1:12-1:13]: Invalid left-hand side in assignment
> 1 | [...x, y] = [1,2,3];
    |             ^ Invalid left-hand side in assignment"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > [...x,] = [1,2,3]; 1`] = `
"SyntaxError [1:10-1:11]: Invalid left-hand side in assignment
> 1 | [...x,] = [1,2,3];
    |           ^ Invalid left-hand side in assignment"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function f(a, ...[b]) { "use strict"; } 1`] = `
"SyntaxError [1:24-1:37]: Illegal 'use strict' directive in function with non-simple parameter list
> 1 | function f(a, ...[b]) { "use strict"; }
    |                         ^^^^^^^^^^^^^ Illegal 'use strict' directive in function with non-simple parameter list"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function f(a, ...b) { "use strict"; } 1`] = `
"SyntaxError [1:22-1:35]: Illegal 'use strict' directive in function with non-simple parameter list
> 1 | function f(a, ...b) { "use strict"; }
    |                       ^^^^^^^^^^^^^ Illegal 'use strict' directive in function with non-simple parameter list"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function foo(...[a], ...b) {} 1`] = `
"SyntaxError [1:19-1:20]: The binding declaration is not destructible
> 1 | function foo(...[a], ...b) {}
    |                    ^ The binding declaration is not destructible"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function foo(...[a],) {} 1`] = `
"SyntaxError [1:19-1:20]: The binding declaration is not destructible
> 1 | function foo(...[a],) {}
    |                    ^ The binding declaration is not destructible"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function foo(...a, ...[b]) {} 1`] = `
"SyntaxError [1:17-1:18]: The binding declaration is not destructible
> 1 | function foo(...a, ...[b]) {}
    |                  ^ The binding declaration is not destructible"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function foo(a, ...[b], c) => {} 1`] = `
"SyntaxError [1:22-1:23]: The binding declaration is not destructible
> 1 | function foo(a, ...[b], c) => {}
    |                       ^ The binding declaration is not destructible"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > function foo(a, ...b, c) => {} 1`] = `
"SyntaxError [1:20-1:21]: The binding declaration is not destructible
> 1 | function foo(a, ...b, c) => {}
    |                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > var obj = class { method(a, b = 1, ...c = [2,3]) {} }; 1`] = `
"SyntaxError [1:47-1:48]: The binding declaration is not destructible
> 1 | var obj = class { method(a, b = 1, ...c = [2,3]) {} };
    |                                                ^ The binding declaration is not destructible"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > var x = { set setter(...[x]) {} } 1`] = `
"SyntaxError [1:21-1:24]: Setter function argument must not be a rest parameter
> 1 | var x = { set setter(...[x]) {} }
    |                      ^^^ Setter function argument must not be a rest parameter"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > var x = class { set setter(...[x]) {} } 1`] = `
"SyntaxError [1:27-1:30]: Setter function argument must not be a rest parameter
> 1 | var x = class { set setter(...[x]) {} }
    |                            ^^^ Setter function argument must not be a rest parameter"
`;

exports[`Expressions - Rest > Expressions - Rest (fail) > var x = class { set setter(...x) {} } 1`] = `
"SyntaxError [1:27-1:30]: Setter function argument must not be a rest parameter
> 1 | var x = class { set setter(...x) {} }
    |                            ^^^ Setter function argument must not be a rest parameter"
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > class restClass { method(a, b, c, ...[d]) { arguments; return [a, b, c, d]; } }; 1`] = `
{
  "body": [
    {
      "body": {
        "body": [
          {
            "computed": false,
            "key": {
              "name": "method",
              "type": "Identifier",
            },
            "kind": "method",
            "static": false,
            "type": "MethodDefinition",
            "value": {
              "async": false,
              "body": {
                "body": [
                  {
                    "expression": {
                      "name": "arguments",
                      "type": "Identifier",
                    },
                    "type": "ExpressionStatement",
                  },
                  {
                    "argument": {
                      "elements": [
                        {
                          "name": "a",
                          "type": "Identifier",
                        },
                        {
                          "name": "b",
                          "type": "Identifier",
                        },
                        {
                          "name": "c",
                          "type": "Identifier",
                        },
                        {
                          "name": "d",
                          "type": "Identifier",
                        },
                      ],
                      "type": "ArrayExpression",
                    },
                    "type": "ReturnStatement",
                  },
                ],
                "type": "BlockStatement",
              },
              "generator": false,
              "id": null,
              "params": [
                {
                  "name": "a",
                  "type": "Identifier",
                },
                {
                  "name": "b",
                  "type": "Identifier",
                },
                {
                  "name": "c",
                  "type": "Identifier",
                },
                {
                  "argument": {
                    "elements": [
                      {
                        "name": "d",
                        "type": "Identifier",
                      },
                    ],
                    "type": "ArrayPattern",
                  },
                  "type": "RestElement",
                },
              ],
              "type": "FunctionExpression",
            },
          },
        ],
        "type": "ClassBody",
      },
      "id": {
        "name": "restClass",
        "type": "Identifier",
      },
      "superClass": null,
      "type": "ClassDeclaration",
    },
    {
      "type": "EmptyStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > function foo(a, b, c, ...[d]) { arguments; return [a, b, c, d]; } 1`] = `
{
  "body": [
    {
      "async": false,
      "body": {
        "body": [
          {
            "expression": {
              "name": "arguments",
              "type": "Identifier",
            },
            "type": "ExpressionStatement",
          },
          {
            "argument": {
              "elements": [
                {
                  "name": "a",
                  "type": "Identifier",
                },
                {
                  "name": "b",
                  "type": "Identifier",
                },
                {
                  "name": "c",
                  "type": "Identifier",
                },
                {
                  "name": "d",
                  "type": "Identifier",
                },
              ],
              "type": "ArrayExpression",
            },
            "type": "ReturnStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "foo",
        "type": "Identifier",
      },
      "params": [
        {
          "name": "a",
          "type": "Identifier",
        },
        {
          "name": "b",
          "type": "Identifier",
        },
        {
          "name": "c",
          "type": "Identifier",
        },
        {
          "argument": {
            "elements": [
              {
                "name": "d",
                "type": "Identifier",
              },
            ],
            "type": "ArrayPattern",
          },
          "type": "RestElement",
        },
      ],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > function fooInline(a, b, c, ...rest) { arguments; this; return [a, b, c, ...rest]; } 1`] = `
{
  "body": [
    {
      "async": false,
      "body": {
        "body": [
          {
            "expression": {
              "name": "arguments",
              "type": "Identifier",
            },
            "type": "ExpressionStatement",
          },
          {
            "expression": {
              "type": "ThisExpression",
            },
            "type": "ExpressionStatement",
          },
          {
            "argument": {
              "elements": [
                {
                  "name": "a",
                  "type": "Identifier",
                },
                {
                  "name": "b",
                  "type": "Identifier",
                },
                {
                  "name": "c",
                  "type": "Identifier",
                },
                {
                  "argument": {
                    "name": "rest",
                    "type": "Identifier",
                  },
                  "type": "SpreadElement",
                },
              ],
              "type": "ArrayExpression",
            },
            "type": "ReturnStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "fooInline",
        "type": "Identifier",
      },
      "params": [
        {
          "name": "a",
          "type": "Identifier",
        },
        {
          "name": "b",
          "type": "Identifier",
        },
        {
          "name": "c",
          "type": "Identifier",
        },
        {
          "argument": {
            "name": "rest",
            "type": "Identifier",
          },
          "type": "RestElement",
        },
      ],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > function objRest(...{'0': a, '1': b, length}) { return [a, b, length]; } 1`] = `
{
  "body": [
    {
      "async": false,
      "body": {
        "body": [
          {
            "argument": {
              "elements": [
                {
                  "name": "a",
                  "type": "Identifier",
                },
                {
                  "name": "b",
                  "type": "Identifier",
                },
                {
                  "name": "length",
                  "type": "Identifier",
                },
              ],
              "type": "ArrayExpression",
            },
            "type": "ReturnStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "objRest",
        "type": "Identifier",
      },
      "params": [
        {
          "argument": {
            "properties": [
              {
                "computed": false,
                "key": {
                  "type": "Literal",
                  "value": "0",
                },
                "kind": "init",
                "method": false,
                "shorthand": false,
                "type": "Property",
                "value": {
                  "name": "a",
                  "type": "Identifier",
                },
              },
              {
                "computed": false,
                "key": {
                  "type": "Literal",
                  "value": "1",
                },
                "kind": "init",
                "method": false,
                "shorthand": false,
                "type": "Property",
                "value": {
                  "name": "b",
                  "type": "Identifier",
                },
              },
              {
                "computed": false,
                "key": {
                  "name": "length",
                  "type": "Identifier",
                },
                "kind": "init",
                "method": false,
                "shorthand": true,
                "type": "Property",
                "value": {
                  "name": "length",
                  "type": "Identifier",
                },
              },
            ],
            "type": "ObjectPattern",
          },
          "type": "RestElement",
        },
      ],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > function singleRest(...[d]) { return d; } 1`] = `
{
  "body": [
    {
      "async": false,
      "body": {
        "body": [
          {
            "argument": {
              "name": "d",
              "type": "Identifier",
            },
            "type": "ReturnStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "singleRest",
        "type": "Identifier",
      },
      "params": [
        {
          "argument": {
            "elements": [
              {
                "name": "d",
                "type": "Identifier",
              },
            ],
            "type": "ArrayPattern",
          },
          "type": "RestElement",
        },
      ],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > var func5 = function (...[argArr13]) { function foo() { eval(); } }; 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "func5",
            "type": "Identifier",
          },
          "init": {
            "async": false,
            "body": {
              "body": [
                {
                  "async": false,
                  "body": {
                    "body": [
                      {
                        "expression": {
                          "arguments": [],
                          "callee": {
                            "name": "eval",
                            "type": "Identifier",
                          },
                          "optional": false,
                          "type": "CallExpression",
                        },
                        "type": "ExpressionStatement",
                      },
                    ],
                    "type": "BlockStatement",
                  },
                  "generator": false,
                  "id": {
                    "name": "foo",
                    "type": "Identifier",
                  },
                  "params": [],
                  "type": "FunctionDeclaration",
                },
              ],
              "type": "BlockStatement",
            },
            "generator": false,
            "id": null,
            "params": [
              {
                "argument": {
                  "elements": [
                    {
                      "name": "argArr13",
                      "type": "Identifier",
                    },
                  ],
                  "type": "ArrayPattern",
                },
                "type": "RestElement",
              },
            ],
            "type": "FunctionExpression",
          },
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Rest > Expressions - Rest (pass) > var obj = { method(a, b, c, ...[d]) { return [a, b, c, d]; } }; 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "obj",
            "type": "Identifier",
          },
          "init": {
            "properties": [
              {
                "computed": false,
                "key": {
                  "name": "method",
                  "type": "Identifier",
                },
                "kind": "init",
                "method": true,
                "shorthand": false,
                "type": "Property",
                "value": {
                  "async": false,
                  "body": {
                    "body": [
                      {
                        "argument": {
                          "elements": [
                            {
                              "name": "a",
                              "type": "Identifier",
                            },
                            {
                              "name": "b",
                              "type": "Identifier",
                            },
                            {
                              "name": "c",
                              "type": "Identifier",
                            },
                            {
                              "name": "d",
                              "type": "Identifier",
                            },
                          ],
                          "type": "ArrayExpression",
                        },
                        "type": "ReturnStatement",
                      },
                    ],
                    "type": "BlockStatement",
                  },
                  "generator": false,
                  "id": null,
                  "params": [
                    {
                      "name": "a",
                      "type": "Identifier",
                    },
                    {
                      "name": "b",
                      "type": "Identifier",
                    },
                    {
                      "name": "c",
                      "type": "Identifier",
                    },
                    {
                      "argument": {
                        "elements": [
                          {
                            "name": "d",
                            "type": "Identifier",
                          },
                        ],
                        "type": "ArrayPattern",
                      },
                      "type": "RestElement",
                    },
                  ],
                  "type": "FunctionExpression",
                },
              },
            ],
            "type": "ObjectExpression",
          },
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
